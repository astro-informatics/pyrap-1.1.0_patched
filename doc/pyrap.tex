\section{Introduction}
Since long glish bindings to the \href{http://casacore.googlecode.com}{casacore} system have been in
place. Quite recently Python bindings have been created in the general
casapy framework using tools like CCMTools, Xerces, Xalan, and
IDL. Albeit very flexible, it is quite complicated and it is not
straightforward to build on other systems than RedHat and OS-X.

Therefore an attempt has been made to make a simpler Python binding
using \href{http://www.boost.org/libs/python/doc}{Boost.Python}.
This proved to be very easy and succesful.
The binding consists of two parts:
\begin{itemize}
\item Converters to translate objects between Python and C++.
\item Class wrappers to map a C++ class and its functions to Python.
\end{itemize}
The Python numarray and numpy (version 1.0 or higher) packages are
supported. At build time one can choose which ones should be used.

\section{Converters}
Boost.Python offers a nice way to convert objects to and from Python.
Ralf W. Grosse-Kunstleve {\tt<rwgk@yahoo.com>} of
\href{http://www.lbl.gov}{Lawrence Berkeley National Laboratory}
has built converters for standard STL containers. This has been
extended to convert to/from other objects.
\\The following C++ objects are currently supported:
\begin{itemize}
  \item scalars (bool, integer, real, complex)
  \item \texttt{std::string}
  \item \texttt{casa::String}
  \item \texttt{std::vector<T>}
  \item \texttt{casa::Vector<T>}
  \item \texttt{casa::IPosition} which defines an array's shape or index.
  \item \texttt{casa::Record} which is a \tt{dict}-like object.
  \item \texttt{casa::ValueHolder} which is a kind of \tt{any} object
    holding a scalar, casa::Array or casa::Record.
  \item exceptions (\texttt{casa::IterError} and \texttt{std::exception})
\end{itemize}
These C++ objects can usually be created from several types of Python
objects and are converted to a specific Python object.
\\Note that the all kinds of numpy (and numarray) objects are
handled by the converters. These are:
\begin{itemize}
\item An array is an N-dimensional (N>0) array of a given type
  (bool, int, float, complex, string).
\item A scalar array is a 0-dimensional array constructed like
  \tt{numpy.array(value)}. Such an array cannot be indexed in Python.
\item An array scalar is an element from an array. The converters
  treat them as normal scalars, which very few Python bindings do.
\end{itemize}
The C++ objects map in the following way to Python objects.
\begin{itemize}
\item
A vector or IPosition object is converted to a Python list.
\\It can be constructed from the following Python objects:
\begin{itemize}
  \item scalar
  \item list or tuple
  \item numarray scalar array or 1-dim array
  \item numpy scalar array or 1-dim array
\end{itemize}
Note that a list or tuple of arbitrary objects can be given. For
example, it is possible to get a \texttt{Vector<TableProxy>} from Python.
\item
A casa::Record is mapped to a Python dict.
\item
Every C++ exception is mapped to a Python \texttt{RunTimeError}
exception. However, \texttt{casa::IterError} is special and 
is mapped to an end-of-iteration exception (\texttt{StopIteration}) in Python.
\item
A casa::ValueHolder is a special \href{http://casacore.googlecode.com}{casacore} object that can hold a record or
a scalar value or N-dim array of many types (bool, numeric, string).
It is meant to conceal the actual type which is useful in functions
that can accept a variety of types (like \texttt{getcell}
in the table binding).
\\Converting a ValueHolder to Python creates the appropriate Python
scalar, array, or dict object. When converting from
Python to ValueHolder, the appropriate internal ValueHolder value
is constructed; a list, tuple, and array object are converted to an
\href{http://casacore.googlecode.com}{casacore} array in the ValueHolder.
\end{itemize}
It means there is no direct Array conversion to/from Python. A
ValueHolder object is always needed to do the conversion. Note that
this is a cheap operation, as it uses Array reference semantics.
ValueHolder has functions to convert between types, so one can get
out an Array with the required type.

\subsection{Array conversion to/from numpy and numarray}
\href{http://casacore.googlecode.com}{casacore} arrays are kept in Fortran-order, while Python arrays are kept
in C-order. It was felt that the Python interface should be as 
pythonic as possible. Therefore it was decided that the array axes are
reversed when converting to/from Python.
The values in an IPosition object (describing shape or
position) are also reversed when converting to/from Python.
\\Note that although numarray and numpy have Fortran-array provisions
by setting the appropriate internal strides,
they do not really support them. When adding, for instance, the scalar
value 0 to a Fortran-array, the  result is a transposed version of the
original (which can be a quite expensive operation). Also operating on
a Fortran-array is twice as slow as operating on a C-array.

A function binding could be such that shape information is passed via, say, a
\texttt{Record} and not via an \texttt{IPosition} object.
In that case its values
are not reversed automatically, so the programmer is
responsible for doing it.

An \href{http://casacore.googlecode.com}{casacore} array is returned to Python as an array object
containing a copy of the \href{http://casacore.googlecode.com}{casacore} array data. If pyrap has
been built with support for only one Python array package (numpy or numarray),
it is clear which array type is returned. If support for both
packages has been built in, by default an array of the imported
package is returned. If both or no array packages have been imported, a numpy
array is returned.
\\Note that there is no support for the old Numeric package.

An \href{http://casacore.googlecode.com}{casacore} array constructed from a Python array is regarded as a
temporary object. So if possible, the \href{http://casacore.googlecode.com}{casacore} array refers to the
Python array data to avoid a needless copy. This is not possible if
the element size in Python differs from \href{http://casacore.googlecode.com}{casacore}. It is also not
possible if the Python array is not contiguous (or not aligned or
byte swapped). In those cases a copy is made.

A few more numarray/numpy specific issues are dealt with:
\begin{itemize}
\item An empty N-dim \href{http://casacore.googlecode.com}{casacore} array (i.e. an array containing no elements) is
returned as an empty N-dim Python array. If the dimensionality is
zero, it is returned as an empty 1-dim array, to prevent numarray/numpy
from treating it as a scalar value.
\item In numarray \texttt{array()} results in \texttt{Py\_None}. This
is accepted by the converters as an empty 1-dim array.
\item Empty arrays can be constructed in Python using empty lists. For
example, \texttt{array([[]])} results in an empty 2-dim array. The
converters accept such empty N-dim Python arrays. The type of an empty
array is set to Int by numarray and to Double by numpy.
\item Because the type of an empty Python array cannot easily be set,
the converters can convert an empty integer or real array to any type.
\item The converters accept a numpy string array. However, it is
returned to Python as the special \texttt{dict} object described
above.
\item Numpy array scalars (elements in an array) are treated by the
converters as ordinary scalar values.
\end{itemize}

\section{Class wrappers}
Usually a binding to an existing Proxy class is made, for example
\texttt{TableProxy}, which should be the same class used in the
glish-binding. 
For a simple binding, only some simple C++ code has to be written in
pyrap\_xx/src/pyxx.cc, where XX is the name of the package/class.
\begin{verbatim}
// Include files for converters being used.
#include <pyrap/Converters/PycExcp.h>
#include <pyrap/Converters/PycBasicData.h>
#include <pyrap/Converters/PycRecord.h>
// Include file for boost python.
#include <boost/python.hpp>

using namespace boost::python;

namespace casa { namespace pyrap {
  void wrap_xx()
  {
    // Define the class; "xx" is the class name in Python.
    class_<XX> ("xx")
      // Define the constructor.
      // Multiple constructors can be defined.
      // They have to have different number of arguments.
      .def (init<>())
      // Add a .def line for each function to be wrapped.
      // An arg line should be added for each argument giving
      // its name and possibly default value.
      .def ("func1", &XX::func1,
            (boost::python::arg("arg1"),
             boost::python::arg("arg2")=0))
    ;
  }
}}

BOOST_PYTHON_MODULE(_xx)
{
  // Register the conversion functions.
  casa::pyrap::register_convert_excp();
  casa::pyrap::register_convert_basicdata();
  casa::pyrap::register_convert_casa_record();
  // Initialize the wrapping.
  casa::pyrap::wrap_xx();
}
\end{verbatim}
Python requires for each package a file \texttt{\_\_init\_\_.py},
so such an empty file should be created as well.

\subsection{More complicated wrappers}
Sometimes a C++ function cannot be wrapped directly, because the argument
order needs to be changed or because some extra Python checks are
necessary.
In such a case the class needs to be implemented in Python itself.
\\The C++ wrapped class name needs to get a different name,
usually by preceeding it with an underscore like:
\begin{verbatim}
    class_<XX> ("_xx")
\end{verbatim}
The Python class should be derived from it and implement the
constructor by calling the constructor of \_xx.
\begin{verbatim}
class xx(_xx):
    def __init__(self):
        _xx.__init__(self)
\end{verbatim}
Now \texttt{xx} inherits all functions from \texttt{\_xx}.
The required function can be written in Python like
\begin{verbatim}
    def func1 (self, arg1, arg2):
        return self._func1 (arg2, arg1);
\end{verbatim}
Note that in the wrapper the function name also needs to be
preceeded by an underscore to make it different.

\subsection{Combining multiple classes}
Sometimes one wants to combine multiple classes in a package. A
example is package \texttt{pyrap\_tables} which contains the classes
\texttt{table}, \texttt{tablecolumn}, \texttt{tablerow},
\texttt{tableiter}, and \texttt{tableindex}. One is referred to the
code of this package to see how
to do it.

\section{Python specifics}
Besides an array being in C-order, there are a few more Python
specific issues.
\begin{itemize}
\item Indexing starts at 0 (vs. 1 in glish).
\item The end value in a range like \texttt{[10:20]} is exclusive
(vs. inclusive in glish). Furthermore Python supports a step and
reversed ranges.
\item Where useful, the function \texttt{\_\_str\_\_} should be added
  giving the name of the object. This function is used when printing
  an object.
\item Where useful, the functions \texttt{\_\_len\_\_},
\texttt{\_\_setitem\_\_(index, value)}, and 
\texttt{\_\_getitem\_\_(index)} should
be added to make it possible that a user indexes an object directly like
\texttt{tabcol[i]} or \texttt{tabcol[start:stop:step]}.
\item When these functions are added, Python supports iteration in an
object. Explicit iteration can also be done by adding the functions
\texttt{\_\_iter\_\_} and \texttt{next}. At the end \texttt{next}
should raise the Python \texttt{StopIteration} exception
(or throw \texttt{casa::IterError}
when implemented in C++) to stop the iteration.
\end{itemize}
