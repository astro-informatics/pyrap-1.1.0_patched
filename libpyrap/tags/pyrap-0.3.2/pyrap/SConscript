import os

Import('env')
lenv =env.Clone()
# to find package based includes
lenv.PrependUnique(CPPPATH=['#'])
lenv.AppendUnique(CPPFLAGS=["-W", "-Wall", "-Woverloaded-virtual"])
buildtype = lenv.get('build_type')
if buildtype == "dbg":
    lenv.AppendUnique(CPPFLAGS=["-g"])
elif buildtype == "opt":
    lenv.AppendUnique(CPPFLAGS=["-O2"])

cpps = lenv.Glob('*/*.cc')
headers = lenv.Glob('*.h')+lenv.Glob('*/*.h')+lenv.Glob('*/*.tcc')

target = 'pyrap'
lib = os.path.join(str(lenv['BUILDDIR']), target)

prefix = lenv.get('prefix')
incdir =  lenv.get('incdir')
libdir =  lenv.get('libdir')
if not incdir:
    incdir = os.path.join(prefix, 'include')
if not libdir:
    libdir = os.path.join(prefix, env["ARCHLIBDIR"])
lenv.Alias('install', incdir)
lenv.Alias('install', libdir)

if lenv.get("enable_static"):
    slib =  lenv.StaticLibrary(target = lib, source = [cpps])
    lenv.Alias(target, slib)
    lenv.Install(libdir, slib)

if not lenv.get("disable_shared"):
    if lenv.get("enable_rpath") and lenv["PLATFORM"] == "darwin":
        name = os.path.join(libdir, '$TARGET.file')
        lenv.PrependUnique(SHLINKFLAGS=["-Wl,-install_name,%s" % name])
    dlib =  lenv.SharedLibrary(target = lib, source = [cpps])
    lenv.Alias(target, dlib)
    lenv.Install(libdir, dlib)

for h in headers:
    lenv.InstallAs(os.path.join(incdir, h.srcnode().path), h)

lenv.Default(target)

testenv = lenv.Clone()
testenv.PrependUnique(LIBPATH=[lenv['BUILDDIR']])
# gather test files
tests = testenv.Glob("*/test/*.py", strings=True)
# allow either linking against shared or static, static if both present
testenv.PrependUnique(LIBS=[target])
for t in tests:
    testenv.addAssayTest(t, alias='test_'+target)
