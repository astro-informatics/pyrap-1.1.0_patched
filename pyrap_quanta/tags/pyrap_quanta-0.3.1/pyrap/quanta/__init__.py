from quantity import quantity, is_quantity
from _quanta import *
constants = constants()
del Quantity, QuantVec, from_string, from_dict_v
