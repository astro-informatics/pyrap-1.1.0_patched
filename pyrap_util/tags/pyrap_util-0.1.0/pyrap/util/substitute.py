# substitute.py: substitute glish variables and expressions
# Copyright (C) 1998,1999,2008
# Associated Universities, Inc. Washington DC, USA.
#
# This library is free software; you can redistribute it and/or modify it
# under the terms of the GNU Library General Public License as published by
# the Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
# License for more details.
#
# You should have received a copy of the GNU Library General Public License
# along with this library; if not, write to the Free Software Foundation,
# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
#
# Correspondence concerning AIPS++ should be addressed as follows:
#        Internet email: aips2-request@nrao.edu.
#        Postal address: AIPS++ Project Office
#                        National Radio Astronomy Observatory
#                        520 Edgemont Road
#                        Charlottesville, VA 22903-2475 USA
#
# $Id: substitute.py 226 2009-11-12 03:39:30Z Malte.Marquarding $
__all__ = ['getlocals', 'getvariable', 'substitute']


def getlocals(back=2):
    """Get the local variables some levels back (-1 is top)."""
    import inspect
    fr = inspect.currentframe()
    try:
        while fr and back != 0:
            fr1 = fr;
            fr = fr.f_back
            back -= 1
    except:
        pass
    return fr1.f_locals

def getvariable(name):
    """Get the value of a local variable somewhere in the call stack"""
    import inspect
    fr = inspect.currentframe()
    try:
        while fr:
            fr  = fr.f_back
            vars = fr.f_locals
            if vars.has_key(name):
                return vars[name]
    except:
        pass
    return None

def substitute(s, objlist=(), globals={}, locals={}):
    """Substitute global python variables in a command string.

    This function parses a string and tries to substitute parts like
    `$name` by their value. It is uses by :mod:`image` and :mod:`table`
    to handle image and table objects in a command, but also other
    variables (integers, strings, etc.) can be substituted.
    The following rules apply:

    1. A name must start with an underscore or alphabetic, followed
       by zero or more alphanumerics and underscores.
    2. String parts enclosed in single or double quotes are literals and
       are left untouched.
       Furthermore a $ can be escaped by a backslash, which is useful
       if an environment variable is used. Note that an extra backslash
       is required in Python to escape the backslash.
       The output contains the quotes and backslashes.
    3. A variable is looked up in the global namespace; that is, in the
       outermost namespace.
    4. If the variable `name` has a vector value, its substitution is
       enclosed in square brackets and separated by commas.
    5. A string value is enclosed in double quotes. If the value
       contains a double quote, that quote is enclosed in single quotes.
    6. If the name's value has a type mentioned in the argument `objlist`,
       it is substituted by `$n` (where n is a sequence number) and its
       value is added to the objects of that type in `objlist`.
    7. If the name is unknown or has an unknown type, it is left untouched.

    The `objlist` argument is a list of tuples or lists where each tuple
    or list has three fields:

    1. The first field is the object type (e.g. `table`)
    2. The second field is a prefix for the sequence number (usually empty).
       E.g. regions could have prefix 'r' resulting in a substitution like `$r1`.
    3. The third field is a list of objects to be substituted. New objects
       get appended to it. Usually the list is initially empty.

    Apart from substituting variables, it also substitutes `$(expression)`
    by the expression result.
    It correctly handles parentheses and quotes in the expression.
    For example::

      a=2
      b=3
      substitute('$(a+b)+$a')              # results in '5+2' (not '7')
      substitute('$((a+b)*(a+b))')         # results in '25'
      substitute('$(len("ab cd( de"))')    # results in '9'

    Substitution is NOT recursive. E.g. if a=1 and b="$a",
    the result of substitute("$b") is "$a" and not 1.

    """
    # Split the string into its individual characters.
    # Initialize some variables.
    backslash = False
    dollar = False
    nparen = 0
    name = ''
    evalstr = ''
    squote = False
    dquote = False
    out = ''
    # Loop through the entire string.
    for tmp in s:
    # If a dollar was found, we might have a name.
    # Alphabetics and underscore are always part of name.
        if dollar:
            if tmp=='_' or (tmp>='a' and tmp<='z') or (tmp>='A' and tmp<='Z'):
                name += tmp
                tmp = ''
            else:
                # Numerics are only part if not first character.
                if tmp>='0' and tmp<='9' and name!='':
                    name += tmp
                    tmp = ''
                else:
                    if tmp=='(' and name=='':
                        # $( indicates the start of a subexpression to evaluate.
                        nparen = 1
                        evalstr = ''
                        tmp = ''
                        dollar = False
                    else:
                        # End of name found. Try to substitute.
                        dollar = False
                        out += substitutename(name, objlist, globals, locals)

        if tmp != '':
            # Handle possible single or double quotes.
            if tmp == '"'  and  not squote:
                dquote = not dquote
            else:
                if tmp == "'"  and  not dquote:
                    squote = not squote
                else:
                    if not dquote and not squote:
                        # Count the number of balanced parentheses
                        # (outside quoted strings)
                        # in the subexpression.
                        if nparen > 0:
                            if tmp == '(':
                                nparen += 1
                            else:
                                if tmp == ')':
                                    nparen -= 1
                                    if nparen == 0:
                                        # The last closing parenthese is found.
                                        # Evaluate the subexpression and if
                                        # successful  put the result in the
                                        # output.
                                        out += substituteexpr(evalstr, globals,
                                                              locals)
                                        tmp = ''
                        else:
                            # Set a switch if we have a dollar (outside quoted
                            # and eval strings)
                            # that is not preceeded by a backslash.
                            if tmp == '$'  and  not backslash:
                                dollar = True
                                name = ''
                                tmp = ''
        # Add the character to output or eval string.
        # Set a switch if we have a backslash.
        if tmp != '':
            if nparen > 0:
                evalstr += tmp
            else:
                out += tmp
        backslash = (tmp == '\\')

        # The entire string has been handled.
        # Substitute a possible last name.
        # Insert a possible incomplete eval string as such.
    if dollar:
        out += substitutename(name, objlist, globals, locals)
    else:
        if nparen > 0:
            out += '$(' + evalstr
    return out


# This function tries to substitute the given name using
# the rules described in the description of function substitute.
def substitutename(name, objlist, globals, locals):
    # If the name is empty, return a single dollar.
    if len(name) == 0:
        return '$'

    # If the name is undefined, return the original.
    try:
        v = eval(name, globals, locals)
#        v = getvariable (name)
    except NameError:
        return '$' + name

    # See if the resulting value is one of the given special types.
    try:
        for objtype,objstr,objs in objlist:
            if isinstance(v, objtype):
                objs += [v]
                return '$' + objstr + str(len(objs))
    except:
        pass

    # No specific type, thus a normal value has to be substituted.
    return substitutevar(v)




# This function tries to substitute the given name using
# the rules described in the description of function substitute.
def substituteexpr(expr, globals={}, locals={}):
    try:
        res = eval(expr, globals, locals)
        v = substitutevar(res)
    except:
        # If the expr is undefined, return the original.
        v = '$(' + expr + ')'
    return str(v)



# Substitute a value.
def substitutevar(v):
    out=''
    if isinstance (v, tuple) or isinstance (v, list):
        out = '['
        first = True
        for tmp in v:
            if first:
                first = False
            else:
                out += ','
            out += substituteonevar(tmp)
        out += ']'
    else:
        out = substituteonevar(v)
    return out


def substituteonevar(v):
    # A string needs to be enclosed in quotes.
    # A vector value is enclosed in square brackets and separated by commas.
    if isinstance (v, str):
        return substitutestring (v)
    # A numeric or boolean value is converted to a string.
    # A vector value is enclosed in square brackets and separated by commas.
    # Take care we have enough precision.
    if isinstance (v, bool):
        if v:
            return 'T'
        return 'F'
    return str(v)


# Enclose a string in double quotes.
# If the string contains double quotes, enclose them in single quotes.
# E.g.                         ab"cd
# is returned as     "ab"'"'"cd"
# which is according to the TaQL rules for strings.
def substitutestring(value):
    out='"'
    for tmp in value:
        if tmp == '"':
            out += '"' + "'" + '"' + "'" + '"'
        else:
            out += tmp
    return out + '"'
