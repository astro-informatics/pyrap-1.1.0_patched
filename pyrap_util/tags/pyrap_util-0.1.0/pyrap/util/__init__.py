"""
Utilities for pyrap modules.
"""
from substitute import substitute, getlocals, getvariable
