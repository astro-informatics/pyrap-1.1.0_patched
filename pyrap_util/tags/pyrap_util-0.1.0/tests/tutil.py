#!/usr/bin/env python

from pyrap.util import substitute

def f1(arg):
    a=3
    s = substitute ('subs as $a $arg', locals=locals())
    print a, arg, s

a=1
b=2
p = "$((a+b)*(a+b))"
s = substitute(p, locals=locals())
print "a=%d, b=%d, %s => %s" % (a, b, p, s) 

f1(23)
f1('xyz')
