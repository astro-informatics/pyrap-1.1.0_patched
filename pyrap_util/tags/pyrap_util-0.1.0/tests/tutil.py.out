a=1, b=2, $((a+b)*(a+b)) => 9
3 23 subs as 3 23
3 xyz subs as 3 "xyz"
